src/ServiceControl/Recoverability/Editing/EditHandler.cs

```diff
@@ -0,0 +1,123 @@
+﻿namespace ServiceControl.Recoverability.Editing
+{
+    using System;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using MessageFailures;
+    using MessageRedirects;
+    using NServiceBus;
+    using NServiceBus.Extensibility;
+    using NServiceBus.Logging;
+    using NServiceBus.Routing;
+    using NServiceBus.Support;
+    using NServiceBus.Transport;
+    using Raven.Abstractions.Data;
+    using Raven.Client;
+
+    class EditHandler : IHandleMessages<EditAndSend>
+    {
+        public EditHandler(IDocumentStore store, IDispatchMessages dispatcher)
+        {
+            this.store = store;
+            this.dispatcher = dispatcher;
+            corruptedReplyToHeaderStrategy = new CorruptedReplyToHeaderStrategy(RuntimeEnvironment.MachineName);
+        }
+
+        public async Task Handle(EditAndSend message, IMessageHandlerContext context)
+        {
+            FailedMessage failedMessage;
+            MessageRedirectsCollection redirects;
+            using (var session = store.OpenAsyncSession())
+            {
+                failedMessage = await session.LoadAsync<FailedMessage>(FailedMessage.MakeDocumentId(message.FailedMessageId))
+                    .ConfigureAwait(false);
+
+                if (failedMessage == null)
+                {
+                    log.WarnFormat("Discarding edit {0} because no message failure for id {1} has been found.", context.MessageId, message.FailedMessageId);
+                    return;
+                }
+
+                var edit = await session.LoadAsync<FailedMessageEdit>(FailedMessageEdit.MakeDocumentId(message.FailedMessageId))
+                    .ConfigureAwait(false);
+                if (edit == null)
+                {
+                    if (failedMessage.Status != FailedMessageStatus.Unresolved)
+                    {
+                        log.WarnFormat("Discarding edit {0} because message failure {1} doesn't have state 'Unresolved'.", context.MessageId, message.FailedMessageId);
+                        return;
+                    }
+
+                    // create a retries document to prevent concurrent edits
+                    await session.StoreAsync(new FailedMessageEdit
+                    {
+                        Id = FailedMessageEdit.MakeDocumentId(message.FailedMessageId),
+                        FailedMessageId = message.FailedMessageId,
+                        EditId = context.MessageId
+                    }, Etag.Empty).ConfigureAwait(false);
```

Comment:
I'm not sure the Etag.Empty approach indeed works. The goal was to make sure that it's not possible to overwrite an existing document (as it would have a different etag). Mabye Raven experts can verify that.

Reply:
Yep, specifying Etag.Empty is how you go about getting a concurrency exception if you're trying to create a new document but someone beat you to it.